File: works/01_bch_osd_reproduction/code/python/experiments/tables.py
"""Tables I–IV reproduction: decoding complexity and latency.

Table I  ((63, 45) at Eb/N0 ∈ {4, 5, 6} dB) — OSD(1), LLOSD(3), LLOSD-B(3)
Table II ((63, 45) at Eb/N0 ∈ {4, 5, 6} dB) — LLOSD-B(3), SLLOSD-B(3,2), YSVL OSD(1), CJ OSD(1)
Table III ((127, 99) at Eb/N0 ∈ {4, 5, 6} dB) — OSD(1), LLOSD(3), HSD(1,8), HSD(1,6)
Table IV  ((255, 223) at Eb/N0 ∈ {4, 5, 6} dB) — OSD(1), LLOSD(2), PLCC(8), HSD(1,8)

Each table has (A) Decoding Complexity (F_2, F_2^m, floating ops) and (B)
Decoding Latency (μs).  We instrument each decoder and average over MC frames.
"""
from work_paths import setup
setup()

import json
import numpy as np

from src.bch import BCHCode, bpsk_modulate, sigma_from_ebn0
from src.osd import osd_decode
from src.decoders import llosd_fast, sllosd_fast, hsd_fast
from src.baselines import ysvl_osd_decode, cj_osd_decode, plcc_decode


def instrument(code, decoder_fn, ebn0, n_trials=500, seed=0):
    """Return avg (f2, f2m, fp, latency_us) over n_trials frames."""
    rate = code.k / code.n
    zero = np.zeros(code.n, dtype=np.int8)
    x0 = bpsk_modulate(zero)
    rng = np.random.default_rng(seed + int(round(ebn0 * 100)))
    sigma = sigma_from_ebn0(ebn0, rate)
    f2, f2m, fp, lat = 0.0, 0.0, 0.0, 0.0
    for _ in range(n_trials):
        y = x0 + sigma * rng.standard_normal(code.n)
        L = 2.0 * y / (sigma * sigma)
        _, s = decoder_fn(code, L)
        c = s["counters"]
        f2  += c.f2
        f2m += c.f2m
        fp  += c.fp
        lat += c.latency_us
    return (f2 / n_trials, f2m / n_trials, fp / n_trials, lat / n_trials)


def build_table(code, decoders, ebn0s, n_trials=500):
    """decoders: list of (name, decoder_fn). Returns dict of dicts."""
    tbl = {}
    for name, fn in decoders:
        row = {}
        for ebn0 in ebn0s:
            f2, f2m, fp, lat = instrument(code, fn, ebn0, n_trials=n_trials)
            row[f"{ebn0}dB"] = {"F2": f2, "F2m": f2m, "FP": fp, "latency_us": lat}
            print(f"  {name} @ {ebn0} dB: F2={f2:.2e}  F2m={f2m:.2e}  FP={fp:.2f}  lat={lat:.1f}μs")
        tbl[name] = row
    return tbl


def format_scientific(x):
    if x < 1:
        return f"{x:.2f}"
    return f"{x:.2e}"


def dump_markdown(name, tbl, ebn0s, path):
    lines = [f"# {name}\n"]
    lines.append("| Algorithm | Eb/N0 (dB) | F_2 ops | F_2^m ops | Floating ops | Latency (μs) |")
    lines.append("|---|---|---|---|---|---|")
    for algo, row in tbl.items():
        first = True
        for ebn0 in ebn0s:
            r = row[f"{ebn0}dB"]
            name_col = algo if first else ""
            lines.append(f"| {name_col} | {ebn0} | {format_scientific(r['F2'])} | "
                          f"{format_scientific(r['F2m'])} | {r['FP']:.1f} | {r['latency_us']:.1f} |")
            first = False
    with open(path, "w") as f:
        f.write("\n".join(lines) + "\n")


def main():
    ebn0s = [4, 5, 6]

    # Table I
    print("=== Table I: (63, 45) ===")
    code = BCHCode(m=6, t=3)
    decoders = [
        ("OSD(1)", lambda c, L: osd_decode(c, L, tau=1)),
        ("LLOSD(3)", lambda c, L: llosd_fast(c, L, tau=3, use_binary_reencoding=False)),
        ("LLOSD-B(3)", lambda c, L: llosd_fast(c, L, tau=3, use_binary_reencoding=True)),
    ]
    tbl1 = build_table(code, decoders, ebn0s, n_trials=500)
    with open("tables/table_I.json", "w") as f:
        json.dump(tbl1, f, indent=2)
    dump_markdown("Table I. Numerical Results in Decoding the (63, 45) BCH Code",
                  tbl1, ebn0s, "tables/table_I.md")

    # Table II
    print("\n=== Table II: (63, 45) — with YSVL / CJ ===")
    decoders2 = [
        ("LLOSD-B(3)", lambda c, L: llosd_fast(c, L, tau=3, use_binary_reencoding=True)),
        ("SLLOSD-B(3,2)", lambda c, L: sllosd_fast(c, L, theta_tuple=(3, 2), use_binary_reencoding=True)),
        ("YSVL OSD(1)", lambda c, L: ysvl_osd_decode(c, L, tau=1)),
        ("CJ OSD(1)", lambda c, L: cj_osd_decode(c, L, tau=1)),
    ]
    tbl2 = build_table(code, decoders2, ebn0s, n_trials=500)
    with open("tables/table_II.json", "w") as f:
        json.dump(tbl2, f, indent=2)
    dump_markdown("Table II. Numerical Results in Decoding the (63, 45) BCH Code",
                  tbl2, ebn0s, "tables/table_II.md")

    # Table III
    print("\n=== Table III: (127, 99) ===")
    code = BCHCode(m=7, t=4)
    decoders3 = [
        ("OSD(1)", lambda c, L: osd_decode(c, L, tau=1)),
        ("LLOSD(3)", lambda c, L: llosd_fast(c, L, tau=3, use_binary_reencoding=True)),
        ("HSD(1,8)", lambda c, L: hsd_fast(c, L, tau=1, eta=8)),
        ("HSD(1,6)", lambda c, L: hsd_fast(c, L, tau=1, eta=6)),
    ]
    tbl3 = build_table(code, decoders3, ebn0s, n_trials=200)
    with open("tables/table_III.json", "w") as f:
        json.dump(tbl3, f, indent=2)
    dump_markdown("Table III. Numerical Results in Decoding the (127, 99) BCH Code",
                  tbl3, ebn0s, "tables/table_III.md")

    # Table IV
    print("\n=== Table IV: (255, 223) ===")
    code = BCHCode(m=8, t=4)
    decoders4 = [
        ("OSD(1)", lambda c, L: osd_decode(c, L, tau=1)),
        ("LLOSD(2)", lambda c, L: llosd_fast(c, L, tau=2, use_binary_reencoding=True)),
        ("PLCC(8)", lambda c, L: plcc_decode(c, L, eta=8)),
        ("HSD(1,8)", lambda c, L: hsd_fast(c, L, tau=1, eta=8)),
    ]
    tbl4 = build_table(code, decoders4, ebn0s, n_trials=100)
    with open("tables/table_IV.json", "w") as f:
        json.dump(tbl4, f, indent=2)
    dump_markdown("Table IV. Numerical Results in Decoding the (255, 223) BCH Code",
                  tbl4, ebn0s, "tables/table_IV.md")


if __name__ == "__main__":
    main()

File: works/01_bch_osd_reproduction/code/matlab/experiments/tables.m
function allTables = tables(options)
%TABLES Reproduce Tables I-IV, including JSON and Markdown output.
% `n_trials_scale` is provided only for an intentionally short run.

if nargin<1, options=struct(); end
paths=work01_setup(); scale=local_opt(options,'n_trials_scale',1); ebn0s=[4,5,6];

code63=work01.Core.bch_code(6,3);
allTables=struct();
fprintf('=== Table I: (63,45) ===\n');
allTables.table_I=local_build(code63,{ ...
    'OSD(1)',@(c,L)work01.Core.osd_decode(c,L,1); ...
    'LLOSD(3)',@(c,L)work01.Core.llosd_fast(c,L,3,false); ...
    'LLOSD-B(3)',@(c,L)work01.Core.llosd_fast(c,L,3,true)},ebn0s,local_n(500,scale));
local_write_table(paths,'table_I',allTables.table_I,ebn0s,'Table I. Numerical Results in Decoding the (63,45) BCH Code');

fprintf('=== Table II: (63,45) ===\n');
allTables.table_II=local_build(code63,{ ...
    'LLOSD-B(3)',@(c,L)work01.Core.llosd_fast(c,L,3,true); ...
    'SLLOSD-B(3,2)',@(c,L)work01.Core.sllosd_fast(c,L,[3,2],true); ...
    'YSVL OSD(1)',@(c,L)work01.Core.ysvl_osd_decode(c,L,1); ...
    'CJ OSD(1)',@(c,L)work01.Core.cj_osd_decode(c,L,1)},ebn0s,local_n(500,scale));
local_write_table(paths,'table_II',allTables.table_II,ebn0s,'Table II. Numerical Results in Decoding the (63,45) BCH Code');

code127=work01.Core.bch_code(7,4); fprintf('=== Table III: (127,99) ===\n');
allTables.table_III=local_build(code127,{ ...
    'OSD(1)',@(c,L)work01.Core.osd_decode(c,L,1); ...
    'LLOSD(3)',@(c,L)work01.Core.llosd_fast(c,L,3,true); ...
    'HSD(1,8)',@(c,L)work01.Core.hsd_fast(c,L,1,8); ...
    'HSD(1,6)',@(c,L)work01.Core.hsd_fast(c,L,1,6)},ebn0s,local_n(200,scale));
local_write_table(paths,'table_III',allTables.table_III,ebn0s,'Table III. Numerical Results in Decoding the (127,99) BCH Code');

code255=work01.Core.bch_code(8,4); fprintf('=== Table IV: (255,223) ===\n');
allTables.table_IV=local_build(code255,{ ...
    'OSD(1)',@(c,L)work01.Core.osd_decode(c,L,1); ...
    'LLOSD(2)',@(c,L)work01.Core.llosd_fast(c,L,2,true); ...
    'PLCC(8)',@(c,L)work01.Core.plcc_decode(c,L,8); ...
    'HSD(1,8)',@(c,L)work01.Core.hsd_fast(c,L,1,8)},ebn0s,local_n(100,scale));
local_write_table(paths,'table_IV',allTables.table_IV,ebn0s,'Table IV. Numerical Results in Decoding the (255,223) BCH Code');
end

function tbl=local_build(code,pairs,ebn0s,nTrials)
tbl=containers.Map;
for i=1:size(pairs,1)
    name=pairs{i,1}; fn=pairs{i,2}; row=containers.Map;
    for ebn0=ebn0s
        metrics=work01.Experiment.instrument(code,fn,ebn0,nTrials,0);
        row(sprintf('%ddB',ebn0))=metrics;
        fprintf('  %s @ %d dB: F2=%.2e F2m=%.2e FP=%.1f lat=%.1f us\n',name,ebn0,metrics.F2,metrics.F2m,metrics.FP,metrics.latency_us);
    end
    tbl(name)=row;
end
end

function local_write_table(paths,name,tbl,ebn0s,titleText)
write_json_file(fullfile(paths.tables_dir,[name '.json']),tbl);
fid=fopen(fullfile(paths.tables_dir,[name '.md']),'w','n','UTF-8'); cleanup=onCleanup(@()fclose(fid)); %#ok<NASGU>
fprintf(fid,'# %s\n\n',titleText); fprintf(fid,'| Algorithm | Eb/N0 (dB) | F_2 ops | F_2^m ops | Floating ops | Latency (us) |\n'); fprintf(fid,'|---|---|---|---|---|---|\n');
names=tbl.keys;
for i=1:numel(names)
    row=tbl(names{i});
    for j=1:numel(ebn0s)
        metric=row(sprintf('%ddB',ebn0s(j))); algo=''; if j==1,algo=names{i};end
        fprintf(fid,'| %s | %d | %s | %s | %.1f | %.1f |\n',algo,ebn0s(j),local_sci(metric.F2),local_sci(metric.F2m),metric.FP,metric.latency_us);
    end
end
end

function s=local_sci(x), if x<1,s=sprintf('%.2f',x);else,s=sprintf('%.2e',x);end,end
function value=local_opt(s,name,default),if isfield(s,name),value=s.(name);else,value=default;end,end
function n=local_n(base,scale),n=max(1,round(base*scale));end
